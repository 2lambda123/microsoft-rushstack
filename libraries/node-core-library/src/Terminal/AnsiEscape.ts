// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { ConsoleColorCodes } from './Colors';

/**
 * Options for {@link AnsiEscape.formatForTests}.
 * @public
 */
export interface IAnsiEscapeConvertForTestsOptions {
  /**
   * If true then `\n` will be replaced by `[n]`, and `\r` will be replaced by `[r]`.
   */
  encodeNewlines?: boolean;
}

/**
 * Operations for encoding and decoding text strings containing
 * {@link https://en.wikipedia.org/wiki/ANSI_escape_code | ANSI escape codes}.
 * The most commonly used escape codes set the foreground/background color for
 * console output.
 * @public
 */
export class AnsiEscape {
  // For now, we only care about the Control Sequence Introducer (CSI) commands which always start with "[".
  // eslint-disable-next-line no-control-regex
  private static _csiRegExp: RegExp = /\x1b\[([\x30-\x3f]*[\x20-\x2f]*[\x40-\x7e])/gu;

  // Text coloring is performed using Select Graphic Rendition (SGR) codes, which come after the
  // CSI introducer "ESC [".  The SGR sequence is a number followed by "m".
  private static _sgrRegExp: RegExp = /([0-9]+)m/u;

  /**
   * Replaces ANSI escape codes with human-readable tokens.  This is useful for unit tests
   * that compare text strings in test assertions or snapshot files.
   */
  public static formatForTests(text: string, options?: IAnsiEscapeConvertForTestsOptions): string {
    if (!options) {
      options = {};
    }

    let result: string = text.replace(AnsiEscape._csiRegExp, (capture: string, csiCode: string) => {
      const match: RegExpMatchArray | null = csiCode.match(AnsiEscape._sgrRegExp);

      if (match) {
        const sgiCode: number = parseInt(match[1]);
        const colorCode: string | undefined = ConsoleColorCodes[sgiCode];
        if (colorCode) {
          // Example: "[BlackForeground]"
          return `[${colorCode}]`;
        }
      }

      // Example: "[31m]"
      return `[${csiCode}]`;
    });

    if (options.encodeNewlines) {
      result = result.replace('\n', '[n]').replace('\r', `[r]`);
    }
    return result;
  }
}
