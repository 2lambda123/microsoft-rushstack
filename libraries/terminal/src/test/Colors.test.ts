// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { Terminal } from '../Terminal';
import { StringBufferTerminalProvider } from '../StringBufferTerminalProvider';
import { createColorGrid } from './createColorGrid';
import { Colors, type IColorableSequence } from '../Colors';

describe(Colors.name, () => {
  let terminal: Terminal;
  let provider: StringBufferTerminalProvider;

  beforeEach(() => {
    provider = new StringBufferTerminalProvider(true);
    terminal = new Terminal(provider);
  });

  test('writes color grid correctly', () => {
    let lineCount: number = 0;
    for (const line of createColorGrid()) {
      terminal.writeLine(...line);
      const output = provider.getOutput();
      expect(output).toMatchSnapshot(`line ${lineCount++}`);
    }

    expect(lineCount).toMatchInlineSnapshot(`10`);
  });

  it('generates codes as expected', () => {
    type ColorsFunctionNames = {
      [K in keyof typeof Colors]: (typeof Colors)[K] extends (str: string) => IColorableSequence ? K : never;
    }[keyof typeof Colors];
    function testColorFunction(functionName: ColorsFunctionNames): void {
      terminal.write(Colors[functionName]('x'));
      expect(provider.getOutput({ normalizeSpecialCharacters: false })).toMatchSnapshot(functionName);
    }

    testColorFunction('black');
    testColorFunction('red');
    testColorFunction('green');
    testColorFunction('yellow');
    testColorFunction('blue');
    testColorFunction('magenta');
    testColorFunction('cyan');
    testColorFunction('white');
    testColorFunction('gray');
    testColorFunction('blackBackground');
    testColorFunction('redBackground');
    testColorFunction('greenBackground');
    testColorFunction('yellowBackground');
    testColorFunction('blueBackground');
    testColorFunction('magentaBackground');
    testColorFunction('cyanBackground');
    testColorFunction('whiteBackground');
    testColorFunction('grayBackground');
    testColorFunction('bold');
    testColorFunction('dim');
    testColorFunction('underline');
    testColorFunction('blink');
    testColorFunction('invertColor');
    testColorFunction('hidden');
  });
});
