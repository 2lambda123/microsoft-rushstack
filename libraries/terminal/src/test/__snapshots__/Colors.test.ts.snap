// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Colors generates codes as expected: black 1`] = `"[30mx[39m"`;

exports[`Colors generates codes as expected: blackBackground 1`] = `"[30mx[39m[31mx[39m[32mx[39m[33mx[39m[34mx[39m[35mx[39m[36mx[39m[37mx[39m[90mx[39m[40mx[49m"`;

exports[`Colors generates codes as expected: blink 1`] = `"[30mx[39m[31mx[39m[32mx[39m[33mx[39m[34mx[39m[35mx[39m[36mx[39m[37mx[39m[90mx[39m[40mx[49m[41mx[49m[42mx[49m[43mx[49m[44mx[49m[45mx[49m[46mx[49m[47mx[49m[100mx[49m[1mx[22m[2mx[22m[4mx[24m[5mx[25m"`;

exports[`Colors generates codes as expected: blue 1`] = `"[30mx[39m[31mx[39m[32mx[39m[33mx[39m[34mx[39m"`;

exports[`Colors generates codes as expected: blueBackground 1`] = `"[30mx[39m[31mx[39m[32mx[39m[33mx[39m[34mx[39m[35mx[39m[36mx[39m[37mx[39m[90mx[39m[40mx[49m[41mx[49m[42mx[49m[43mx[49m[44mx[49m"`;

exports[`Colors generates codes as expected: bold 1`] = `"[30mx[39m[31mx[39m[32mx[39m[33mx[39m[34mx[39m[35mx[39m[36mx[39m[37mx[39m[90mx[39m[40mx[49m[41mx[49m[42mx[49m[43mx[49m[44mx[49m[45mx[49m[46mx[49m[47mx[49m[100mx[49m[1mx[22m"`;

exports[`Colors generates codes as expected: cyan 1`] = `"[30mx[39m[31mx[39m[32mx[39m[33mx[39m[34mx[39m[35mx[39m[36mx[39m"`;

exports[`Colors generates codes as expected: cyanBackground 1`] = `"[30mx[39m[31mx[39m[32mx[39m[33mx[39m[34mx[39m[35mx[39m[36mx[39m[37mx[39m[90mx[39m[40mx[49m[41mx[49m[42mx[49m[43mx[49m[44mx[49m[45mx[49m[46mx[49m"`;

exports[`Colors generates codes as expected: dim 1`] = `"[30mx[39m[31mx[39m[32mx[39m[33mx[39m[34mx[39m[35mx[39m[36mx[39m[37mx[39m[90mx[39m[40mx[49m[41mx[49m[42mx[49m[43mx[49m[44mx[49m[45mx[49m[46mx[49m[47mx[49m[100mx[49m[1mx[22m[2mx[22m"`;

exports[`Colors generates codes as expected: gray 1`] = `"[30mx[39m[31mx[39m[32mx[39m[33mx[39m[34mx[39m[35mx[39m[36mx[39m[37mx[39m[90mx[39m"`;

exports[`Colors generates codes as expected: grayBackground 1`] = `"[30mx[39m[31mx[39m[32mx[39m[33mx[39m[34mx[39m[35mx[39m[36mx[39m[37mx[39m[90mx[39m[40mx[49m[41mx[49m[42mx[49m[43mx[49m[44mx[49m[45mx[49m[46mx[49m[47mx[49m[100mx[49m"`;

exports[`Colors generates codes as expected: green 1`] = `"[30mx[39m[31mx[39m[32mx[39m"`;

exports[`Colors generates codes as expected: greenBackground 1`] = `"[30mx[39m[31mx[39m[32mx[39m[33mx[39m[34mx[39m[35mx[39m[36mx[39m[37mx[39m[90mx[39m[40mx[49m[41mx[49m[42mx[49m"`;

exports[`Colors generates codes as expected: hidden 1`] = `"[30mx[39m[31mx[39m[32mx[39m[33mx[39m[34mx[39m[35mx[39m[36mx[39m[37mx[39m[90mx[39m[40mx[49m[41mx[49m[42mx[49m[43mx[49m[44mx[49m[45mx[49m[46mx[49m[47mx[49m[100mx[49m[1mx[22m[2mx[22m[4mx[24m[5mx[25m[7mx[27m[8mx[28m"`;

exports[`Colors generates codes as expected: invertColor 1`] = `"[30mx[39m[31mx[39m[32mx[39m[33mx[39m[34mx[39m[35mx[39m[36mx[39m[37mx[39m[90mx[39m[40mx[49m[41mx[49m[42mx[49m[43mx[49m[44mx[49m[45mx[49m[46mx[49m[47mx[49m[100mx[49m[1mx[22m[2mx[22m[4mx[24m[5mx[25m[7mx[27m"`;

exports[`Colors generates codes as expected: magenta 1`] = `"[30mx[39m[31mx[39m[32mx[39m[33mx[39m[34mx[39m[35mx[39m"`;

exports[`Colors generates codes as expected: magentaBackground 1`] = `"[30mx[39m[31mx[39m[32mx[39m[33mx[39m[34mx[39m[35mx[39m[36mx[39m[37mx[39m[90mx[39m[40mx[49m[41mx[49m[42mx[49m[43mx[49m[44mx[49m[45mx[49m"`;

exports[`Colors generates codes as expected: red 1`] = `"[30mx[39m[31mx[39m"`;

exports[`Colors generates codes as expected: redBackground 1`] = `"[30mx[39m[31mx[39m[32mx[39m[33mx[39m[34mx[39m[35mx[39m[36mx[39m[37mx[39m[90mx[39m[40mx[49m[41mx[49m"`;

exports[`Colors generates codes as expected: underline 1`] = `"[30mx[39m[31mx[39m[32mx[39m[33mx[39m[34mx[39m[35mx[39m[36mx[39m[37mx[39m[90mx[39m[40mx[49m[41mx[49m[42mx[49m[43mx[49m[44mx[49m[45mx[49m[46mx[49m[47mx[49m[100mx[49m[1mx[22m[2mx[22m[4mx[24m"`;

exports[`Colors generates codes as expected: white 1`] = `"[30mx[39m[31mx[39m[32mx[39m[33mx[39m[34mx[39m[35mx[39m[36mx[39m[37mx[39m"`;

exports[`Colors generates codes as expected: whiteBackground 1`] = `"[30mx[39m[31mx[39m[32mx[39m[33mx[39m[34mx[39m[35mx[39m[36mx[39m[37mx[39m[90mx[39m[40mx[49m[41mx[49m[42mx[49m[43mx[49m[44mx[49m[45mx[49m[46mx[49m[47mx[49m"`;

exports[`Colors generates codes as expected: yellow 1`] = `"[30mx[39m[31mx[39m[32mx[39m[33mx[39m"`;

exports[`Colors generates codes as expected: yellowBackground 1`] = `"[30mx[39m[31mx[39m[32mx[39m[33mx[39m[34mx[39m[35mx[39m[36mx[39m[37mx[39m[90mx[39m[40mx[49m[41mx[49m[42mx[49m[43mx[49m"`;

exports[`Colors writes color grid correctly: line 0 1`] = `"X[black]X[default][white]X[default][gray]X[default][magenta]X[default][red]X[default][yellow]X[default][green]X[default][cyan]X[default][blue]X[default][n]"`;

exports[`Colors writes color grid correctly: line 1 1`] = `"X[black]X[default][white]X[default][gray]X[default][magenta]X[default][red]X[default][yellow]X[default][green]X[default][cyan]X[default][blue]X[default][n][black-bg]X[default-bg][black][black-bg]X[default-bg][default][white][black-bg]X[default-bg][default][gray][black-bg]X[default-bg][default][magenta][black-bg]X[default-bg][default][red][black-bg]X[default-bg][default][yellow][black-bg]X[default-bg][default][green][black-bg]X[default-bg][default][cyan][black-bg]X[default-bg][default][blue][black-bg]X[default-bg][default][n]"`;

exports[`Colors writes color grid correctly: line 2 1`] = `"X[black]X[default][white]X[default][gray]X[default][magenta]X[default][red]X[default][yellow]X[default][green]X[default][cyan]X[default][blue]X[default][n][black-bg]X[default-bg][black][black-bg]X[default-bg][default][white][black-bg]X[default-bg][default][gray][black-bg]X[default-bg][default][magenta][black-bg]X[default-bg][default][red][black-bg]X[default-bg][default][yellow][black-bg]X[default-bg][default][green][black-bg]X[default-bg][default][cyan][black-bg]X[default-bg][default][blue][black-bg]X[default-bg][default][n][white-bg]X[default-bg][black][white-bg]X[default-bg][default][white][white-bg]X[default-bg][default][gray][white-bg]X[default-bg][default][magenta][white-bg]X[default-bg][default][red][white-bg]X[default-bg][default][yellow][white-bg]X[default-bg][default][green][white-bg]X[default-bg][default][cyan][white-bg]X[default-bg][default][blue][white-bg]X[default-bg][default][n]"`;

exports[`Colors writes color grid correctly: line 3 1`] = `"X[black]X[default][white]X[default][gray]X[default][magenta]X[default][red]X[default][yellow]X[default][green]X[default][cyan]X[default][blue]X[default][n][black-bg]X[default-bg][black][black-bg]X[default-bg][default][white][black-bg]X[default-bg][default][gray][black-bg]X[default-bg][default][magenta][black-bg]X[default-bg][default][red][black-bg]X[default-bg][default][yellow][black-bg]X[default-bg][default][green][black-bg]X[default-bg][default][cyan][black-bg]X[default-bg][default][blue][black-bg]X[default-bg][default][n][white-bg]X[default-bg][black][white-bg]X[default-bg][default][white][white-bg]X[default-bg][default][gray][white-bg]X[default-bg][default][magenta][white-bg]X[default-bg][default][red][white-bg]X[default-bg][default][yellow][white-bg]X[default-bg][default][green][white-bg]X[default-bg][default][cyan][white-bg]X[default-bg][default][blue][white-bg]X[default-bg][default][n][gray-bg]X[default-bg][black][gray-bg]X[default-bg][default][white][gray-bg]X[default-bg][default][gray][gray-bg]X[default-bg][default][magenta][gray-bg]X[default-bg][default][red][gray-bg]X[default-bg][default][yellow][gray-bg]X[default-bg][default][green][gray-bg]X[default-bg][default][cyan][gray-bg]X[default-bg][default][blue][gray-bg]X[default-bg][default][n]"`;

exports[`Colors writes color grid correctly: line 4 1`] = `"X[black]X[default][white]X[default][gray]X[default][magenta]X[default][red]X[default][yellow]X[default][green]X[default][cyan]X[default][blue]X[default][n][black-bg]X[default-bg][black][black-bg]X[default-bg][default][white][black-bg]X[default-bg][default][gray][black-bg]X[default-bg][default][magenta][black-bg]X[default-bg][default][red][black-bg]X[default-bg][default][yellow][black-bg]X[default-bg][default][green][black-bg]X[default-bg][default][cyan][black-bg]X[default-bg][default][blue][black-bg]X[default-bg][default][n][white-bg]X[default-bg][black][white-bg]X[default-bg][default][white][white-bg]X[default-bg][default][gray][white-bg]X[default-bg][default][magenta][white-bg]X[default-bg][default][red][white-bg]X[default-bg][default][yellow][white-bg]X[default-bg][default][green][white-bg]X[default-bg][default][cyan][white-bg]X[default-bg][default][blue][white-bg]X[default-bg][default][n][gray-bg]X[default-bg][black][gray-bg]X[default-bg][default][white][gray-bg]X[default-bg][default][gray][gray-bg]X[default-bg][default][magenta][gray-bg]X[default-bg][default][red][gray-bg]X[default-bg][default][yellow][gray-bg]X[default-bg][default][green][gray-bg]X[default-bg][default][cyan][gray-bg]X[default-bg][default][blue][gray-bg]X[default-bg][default][n][magenta-bg]X[default-bg][black][magenta-bg]X[default-bg][default][white][magenta-bg]X[default-bg][default][gray][magenta-bg]X[default-bg][default][magenta][magenta-bg]X[default-bg][default][red][magenta-bg]X[default-bg][default][yellow][magenta-bg]X[default-bg][default][green][magenta-bg]X[default-bg][default][cyan][magenta-bg]X[default-bg][default][blue][magenta-bg]X[default-bg][default][n]"`;

exports[`Colors writes color grid correctly: line 5 1`] = `"X[black]X[default][white]X[default][gray]X[default][magenta]X[default][red]X[default][yellow]X[default][green]X[default][cyan]X[default][blue]X[default][n][black-bg]X[default-bg][black][black-bg]X[default-bg][default][white][black-bg]X[default-bg][default][gray][black-bg]X[default-bg][default][magenta][black-bg]X[default-bg][default][red][black-bg]X[default-bg][default][yellow][black-bg]X[default-bg][default][green][black-bg]X[default-bg][default][cyan][black-bg]X[default-bg][default][blue][black-bg]X[default-bg][default][n][white-bg]X[default-bg][black][white-bg]X[default-bg][default][white][white-bg]X[default-bg][default][gray][white-bg]X[default-bg][default][magenta][white-bg]X[default-bg][default][red][white-bg]X[default-bg][default][yellow][white-bg]X[default-bg][default][green][white-bg]X[default-bg][default][cyan][white-bg]X[default-bg][default][blue][white-bg]X[default-bg][default][n][gray-bg]X[default-bg][black][gray-bg]X[default-bg][default][white][gray-bg]X[default-bg][default][gray][gray-bg]X[default-bg][default][magenta][gray-bg]X[default-bg][default][red][gray-bg]X[default-bg][default][yellow][gray-bg]X[default-bg][default][green][gray-bg]X[default-bg][default][cyan][gray-bg]X[default-bg][default][blue][gray-bg]X[default-bg][default][n][magenta-bg]X[default-bg][black][magenta-bg]X[default-bg][default][white][magenta-bg]X[default-bg][default][gray][magenta-bg]X[default-bg][default][magenta][magenta-bg]X[default-bg][default][red][magenta-bg]X[default-bg][default][yellow][magenta-bg]X[default-bg][default][green][magenta-bg]X[default-bg][default][cyan][magenta-bg]X[default-bg][default][blue][magenta-bg]X[default-bg][default][n][red-bg]X[default-bg][black][red-bg]X[default-bg][default][white][red-bg]X[default-bg][default][gray][red-bg]X[default-bg][default][magenta][red-bg]X[default-bg][default][red][red-bg]X[default-bg][default][yellow][red-bg]X[default-bg][default][green][red-bg]X[default-bg][default][cyan][red-bg]X[default-bg][default][blue][red-bg]X[default-bg][default][n]"`;

exports[`Colors writes color grid correctly: line 6 1`] = `"X[black]X[default][white]X[default][gray]X[default][magenta]X[default][red]X[default][yellow]X[default][green]X[default][cyan]X[default][blue]X[default][n][black-bg]X[default-bg][black][black-bg]X[default-bg][default][white][black-bg]X[default-bg][default][gray][black-bg]X[default-bg][default][magenta][black-bg]X[default-bg][default][red][black-bg]X[default-bg][default][yellow][black-bg]X[default-bg][default][green][black-bg]X[default-bg][default][cyan][black-bg]X[default-bg][default][blue][black-bg]X[default-bg][default][n][white-bg]X[default-bg][black][white-bg]X[default-bg][default][white][white-bg]X[default-bg][default][gray][white-bg]X[default-bg][default][magenta][white-bg]X[default-bg][default][red][white-bg]X[default-bg][default][yellow][white-bg]X[default-bg][default][green][white-bg]X[default-bg][default][cyan][white-bg]X[default-bg][default][blue][white-bg]X[default-bg][default][n][gray-bg]X[default-bg][black][gray-bg]X[default-bg][default][white][gray-bg]X[default-bg][default][gray][gray-bg]X[default-bg][default][magenta][gray-bg]X[default-bg][default][red][gray-bg]X[default-bg][default][yellow][gray-bg]X[default-bg][default][green][gray-bg]X[default-bg][default][cyan][gray-bg]X[default-bg][default][blue][gray-bg]X[default-bg][default][n][magenta-bg]X[default-bg][black][magenta-bg]X[default-bg][default][white][magenta-bg]X[default-bg][default][gray][magenta-bg]X[default-bg][default][magenta][magenta-bg]X[default-bg][default][red][magenta-bg]X[default-bg][default][yellow][magenta-bg]X[default-bg][default][green][magenta-bg]X[default-bg][default][cyan][magenta-bg]X[default-bg][default][blue][magenta-bg]X[default-bg][default][n][red-bg]X[default-bg][black][red-bg]X[default-bg][default][white][red-bg]X[default-bg][default][gray][red-bg]X[default-bg][default][magenta][red-bg]X[default-bg][default][red][red-bg]X[default-bg][default][yellow][red-bg]X[default-bg][default][green][red-bg]X[default-bg][default][cyan][red-bg]X[default-bg][default][blue][red-bg]X[default-bg][default][n][yellow-bg]X[default-bg][black][yellow-bg]X[default-bg][default][white][yellow-bg]X[default-bg][default][gray][yellow-bg]X[default-bg][default][magenta][yellow-bg]X[default-bg][default][red][yellow-bg]X[default-bg][default][yellow][yellow-bg]X[default-bg][default][green][yellow-bg]X[default-bg][default][cyan][yellow-bg]X[default-bg][default][blue][yellow-bg]X[default-bg][default][n]"`;

exports[`Colors writes color grid correctly: line 7 1`] = `"X[black]X[default][white]X[default][gray]X[default][magenta]X[default][red]X[default][yellow]X[default][green]X[default][cyan]X[default][blue]X[default][n][black-bg]X[default-bg][black][black-bg]X[default-bg][default][white][black-bg]X[default-bg][default][gray][black-bg]X[default-bg][default][magenta][black-bg]X[default-bg][default][red][black-bg]X[default-bg][default][yellow][black-bg]X[default-bg][default][green][black-bg]X[default-bg][default][cyan][black-bg]X[default-bg][default][blue][black-bg]X[default-bg][default][n][white-bg]X[default-bg][black][white-bg]X[default-bg][default][white][white-bg]X[default-bg][default][gray][white-bg]X[default-bg][default][magenta][white-bg]X[default-bg][default][red][white-bg]X[default-bg][default][yellow][white-bg]X[default-bg][default][green][white-bg]X[default-bg][default][cyan][white-bg]X[default-bg][default][blue][white-bg]X[default-bg][default][n][gray-bg]X[default-bg][black][gray-bg]X[default-bg][default][white][gray-bg]X[default-bg][default][gray][gray-bg]X[default-bg][default][magenta][gray-bg]X[default-bg][default][red][gray-bg]X[default-bg][default][yellow][gray-bg]X[default-bg][default][green][gray-bg]X[default-bg][default][cyan][gray-bg]X[default-bg][default][blue][gray-bg]X[default-bg][default][n][magenta-bg]X[default-bg][black][magenta-bg]X[default-bg][default][white][magenta-bg]X[default-bg][default][gray][magenta-bg]X[default-bg][default][magenta][magenta-bg]X[default-bg][default][red][magenta-bg]X[default-bg][default][yellow][magenta-bg]X[default-bg][default][green][magenta-bg]X[default-bg][default][cyan][magenta-bg]X[default-bg][default][blue][magenta-bg]X[default-bg][default][n][red-bg]X[default-bg][black][red-bg]X[default-bg][default][white][red-bg]X[default-bg][default][gray][red-bg]X[default-bg][default][magenta][red-bg]X[default-bg][default][red][red-bg]X[default-bg][default][yellow][red-bg]X[default-bg][default][green][red-bg]X[default-bg][default][cyan][red-bg]X[default-bg][default][blue][red-bg]X[default-bg][default][n][yellow-bg]X[default-bg][black][yellow-bg]X[default-bg][default][white][yellow-bg]X[default-bg][default][gray][yellow-bg]X[default-bg][default][magenta][yellow-bg]X[default-bg][default][red][yellow-bg]X[default-bg][default][yellow][yellow-bg]X[default-bg][default][green][yellow-bg]X[default-bg][default][cyan][yellow-bg]X[default-bg][default][blue][yellow-bg]X[default-bg][default][n][green-bg]X[default-bg][black][green-bg]X[default-bg][default][white][green-bg]X[default-bg][default][gray][green-bg]X[default-bg][default][magenta][green-bg]X[default-bg][default][red][green-bg]X[default-bg][default][yellow][green-bg]X[default-bg][default][green][green-bg]X[default-bg][default][cyan][green-bg]X[default-bg][default][blue][green-bg]X[default-bg][default][n]"`;

exports[`Colors writes color grid correctly: line 8 1`] = `"X[black]X[default][white]X[default][gray]X[default][magenta]X[default][red]X[default][yellow]X[default][green]X[default][cyan]X[default][blue]X[default][n][black-bg]X[default-bg][black][black-bg]X[default-bg][default][white][black-bg]X[default-bg][default][gray][black-bg]X[default-bg][default][magenta][black-bg]X[default-bg][default][red][black-bg]X[default-bg][default][yellow][black-bg]X[default-bg][default][green][black-bg]X[default-bg][default][cyan][black-bg]X[default-bg][default][blue][black-bg]X[default-bg][default][n][white-bg]X[default-bg][black][white-bg]X[default-bg][default][white][white-bg]X[default-bg][default][gray][white-bg]X[default-bg][default][magenta][white-bg]X[default-bg][default][red][white-bg]X[default-bg][default][yellow][white-bg]X[default-bg][default][green][white-bg]X[default-bg][default][cyan][white-bg]X[default-bg][default][blue][white-bg]X[default-bg][default][n][gray-bg]X[default-bg][black][gray-bg]X[default-bg][default][white][gray-bg]X[default-bg][default][gray][gray-bg]X[default-bg][default][magenta][gray-bg]X[default-bg][default][red][gray-bg]X[default-bg][default][yellow][gray-bg]X[default-bg][default][green][gray-bg]X[default-bg][default][cyan][gray-bg]X[default-bg][default][blue][gray-bg]X[default-bg][default][n][magenta-bg]X[default-bg][black][magenta-bg]X[default-bg][default][white][magenta-bg]X[default-bg][default][gray][magenta-bg]X[default-bg][default][magenta][magenta-bg]X[default-bg][default][red][magenta-bg]X[default-bg][default][yellow][magenta-bg]X[default-bg][default][green][magenta-bg]X[default-bg][default][cyan][magenta-bg]X[default-bg][default][blue][magenta-bg]X[default-bg][default][n][red-bg]X[default-bg][black][red-bg]X[default-bg][default][white][red-bg]X[default-bg][default][gray][red-bg]X[default-bg][default][magenta][red-bg]X[default-bg][default][red][red-bg]X[default-bg][default][yellow][red-bg]X[default-bg][default][green][red-bg]X[default-bg][default][cyan][red-bg]X[default-bg][default][blue][red-bg]X[default-bg][default][n][yellow-bg]X[default-bg][black][yellow-bg]X[default-bg][default][white][yellow-bg]X[default-bg][default][gray][yellow-bg]X[default-bg][default][magenta][yellow-bg]X[default-bg][default][red][yellow-bg]X[default-bg][default][yellow][yellow-bg]X[default-bg][default][green][yellow-bg]X[default-bg][default][cyan][yellow-bg]X[default-bg][default][blue][yellow-bg]X[default-bg][default][n][green-bg]X[default-bg][black][green-bg]X[default-bg][default][white][green-bg]X[default-bg][default][gray][green-bg]X[default-bg][default][magenta][green-bg]X[default-bg][default][red][green-bg]X[default-bg][default][yellow][green-bg]X[default-bg][default][green][green-bg]X[default-bg][default][cyan][green-bg]X[default-bg][default][blue][green-bg]X[default-bg][default][n][cyan-bg]X[default-bg][black][cyan-bg]X[default-bg][default][white][cyan-bg]X[default-bg][default][gray][cyan-bg]X[default-bg][default][magenta][cyan-bg]X[default-bg][default][red][cyan-bg]X[default-bg][default][yellow][cyan-bg]X[default-bg][default][green][cyan-bg]X[default-bg][default][cyan][cyan-bg]X[default-bg][default][blue][cyan-bg]X[default-bg][default][n]"`;

exports[`Colors writes color grid correctly: line 9 1`] = `"X[black]X[default][white]X[default][gray]X[default][magenta]X[default][red]X[default][yellow]X[default][green]X[default][cyan]X[default][blue]X[default][n][black-bg]X[default-bg][black][black-bg]X[default-bg][default][white][black-bg]X[default-bg][default][gray][black-bg]X[default-bg][default][magenta][black-bg]X[default-bg][default][red][black-bg]X[default-bg][default][yellow][black-bg]X[default-bg][default][green][black-bg]X[default-bg][default][cyan][black-bg]X[default-bg][default][blue][black-bg]X[default-bg][default][n][white-bg]X[default-bg][black][white-bg]X[default-bg][default][white][white-bg]X[default-bg][default][gray][white-bg]X[default-bg][default][magenta][white-bg]X[default-bg][default][red][white-bg]X[default-bg][default][yellow][white-bg]X[default-bg][default][green][white-bg]X[default-bg][default][cyan][white-bg]X[default-bg][default][blue][white-bg]X[default-bg][default][n][gray-bg]X[default-bg][black][gray-bg]X[default-bg][default][white][gray-bg]X[default-bg][default][gray][gray-bg]X[default-bg][default][magenta][gray-bg]X[default-bg][default][red][gray-bg]X[default-bg][default][yellow][gray-bg]X[default-bg][default][green][gray-bg]X[default-bg][default][cyan][gray-bg]X[default-bg][default][blue][gray-bg]X[default-bg][default][n][magenta-bg]X[default-bg][black][magenta-bg]X[default-bg][default][white][magenta-bg]X[default-bg][default][gray][magenta-bg]X[default-bg][default][magenta][magenta-bg]X[default-bg][default][red][magenta-bg]X[default-bg][default][yellow][magenta-bg]X[default-bg][default][green][magenta-bg]X[default-bg][default][cyan][magenta-bg]X[default-bg][default][blue][magenta-bg]X[default-bg][default][n][red-bg]X[default-bg][black][red-bg]X[default-bg][default][white][red-bg]X[default-bg][default][gray][red-bg]X[default-bg][default][magenta][red-bg]X[default-bg][default][red][red-bg]X[default-bg][default][yellow][red-bg]X[default-bg][default][green][red-bg]X[default-bg][default][cyan][red-bg]X[default-bg][default][blue][red-bg]X[default-bg][default][n][yellow-bg]X[default-bg][black][yellow-bg]X[default-bg][default][white][yellow-bg]X[default-bg][default][gray][yellow-bg]X[default-bg][default][magenta][yellow-bg]X[default-bg][default][red][yellow-bg]X[default-bg][default][yellow][yellow-bg]X[default-bg][default][green][yellow-bg]X[default-bg][default][cyan][yellow-bg]X[default-bg][default][blue][yellow-bg]X[default-bg][default][n][green-bg]X[default-bg][black][green-bg]X[default-bg][default][white][green-bg]X[default-bg][default][gray][green-bg]X[default-bg][default][magenta][green-bg]X[default-bg][default][red][green-bg]X[default-bg][default][yellow][green-bg]X[default-bg][default][green][green-bg]X[default-bg][default][cyan][green-bg]X[default-bg][default][blue][green-bg]X[default-bg][default][n][cyan-bg]X[default-bg][black][cyan-bg]X[default-bg][default][white][cyan-bg]X[default-bg][default][gray][cyan-bg]X[default-bg][default][magenta][cyan-bg]X[default-bg][default][red][cyan-bg]X[default-bg][default][yellow][cyan-bg]X[default-bg][default][green][cyan-bg]X[default-bg][default][cyan][cyan-bg]X[default-bg][default][blue][cyan-bg]X[default-bg][default][n][blue-bg]X[default-bg][black][blue-bg]X[default-bg][default][white][blue-bg]X[default-bg][default][gray][blue-bg]X[default-bg][default][magenta][blue-bg]X[default-bg][default][red][blue-bg]X[default-bg][default][yellow][blue-bg]X[default-bg][default][green][blue-bg]X[default-bg][default][cyan][blue-bg]X[default-bg][default][blue][blue-bg]X[default-bg][default][n]"`;
